File: Medium/79. Word Search/WordSearch.py
'''
LeetCode Medium: 79. Word Search
https://leetcode.com/problems/word-search/

Check WordSearch.cpp for brief solution explanation.
'''
class Solution:
    def bounds(self, r, c) -> bool:
        return 0<=r<self.M and 0<=c<self.N
    def dfs(self, r, c, board, word, idx) -> bool:
        if idx+1>=len(word):
            return True
        cur = board[r][c]
        board[r][c] = '*'
        res = False
        for i in range(4):
            newR, newC = r+self.neighbours[i][0], c+self.neighbours[i][1]
            if self.bounds(newR, newC) and board[newR][newC]==word[idx+1]:
                res|=self.dfs(newR, newC, board, word, idx+1)
            if res:
                return True
        board[r][c] = cur
        return False
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.M, self.N = len(board), len(board[0])
        self.neighbours = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        for i in range(self.M):
            for j in range(self.N):
                if board[i][j]==word[0] and self.dfs(i, j, board, word, 0):
                    return True;
        return False

File: Medium/79. Word Search/WordSearch.java
/*
LeetCode Medium: 79. Word Search
https://leetcode.com/problems/word-search/

Check WordSearch.cpp for brief solution explanation.
*/
class Solution {
    int[][]neighbours = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public boolean bounds(int r, int c, int M, int N){
        return 0<=r&&r<M&&0<=c&&c<N;
    }
    public boolean dfs(int r, int c, int M, int N, char[][] board, String word, int idx){
        if(idx+1>=word.length()){
            return true;
        }
        char cur = board[r][c];
        board[r][c] = '*';
        boolean res = false;
        for(int i = 0;i<4;i++){
            int newR = r+neighbours[i][0], newC = c+neighbours[i][1];
            if(bounds(newR, newC, M, N)&&board[newR][newC]==word.charAt(idx+1)){
                res|=dfs(newR, newC, M, N, board, word, idx+1);
            }
            if(res){
                return true;
            }
        }
        board[r][c] = cur;
        return false;
    }
    public boolean exist(char[][] board, String word) {
        int M = board.length, N = board[0].length;
        for(int i = 0;i<M;i++){
            for(int j = 0;j<N;j++){
                if(board[i][j]==word.charAt(0)&&dfs(i, j, M, N, board, word, 0)){
                    return true;
                }
            }
        }
        return false;
    }
}